pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,402 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials/install"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials/v1alpha1"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubeletconfigv1alpha1 "k8s.io/kubernetes/pkg/kubelet/apis/config/v1alpha1"
+)
+
+const (
+	globalCacheKey = "global"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+
+	apiVersions = map[string]schema.GroupVersion{
+		v1alpha1.SchemeGroupVersion.String(): v1alpha1.SchemeGroupVersion,
+	}
+)
+
+func init() {
+	install.Install(scheme)
+	kubeletconfig.AddToScheme(scheme)
+	kubeletconfigv1alpha1.AddToScheme(scheme)
+}
+
+// RegisterCredentialProviderPlugins is called from kubelet to register external credential provider
+// plugins according to the CredentialProviderConfig config file.
+func RegisterCredentialProviderPlugins(pluginConfigFile, pluginBinDir string) error {
+	credentialProviderConfig, err := readCredentialProviderConfigFile(pluginConfigFile)
+	if err != nil {
+		return err
+	}
+
+	errs := validateCredentialProviderConfig(credentialProviderConfig)
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to validate credential provider config: %v", errs.ToAggregate())
+	}
+
+	for _, provider := range credentialProviderConfig.Providers {
+		plugin, err := newPluginProvider(pluginBinDir, provider)
+		if err != nil {
+			return fmt.Errorf("error initializing plugin provider %s: %w", provider.Name, err)
+		}
+
+		credentialprovider.RegisterCredentialProvider(provider.Name, plugin)
+	}
+
+	return nil
+}
+
+// newPluginProvider returns a new pluginProvider based on the credential provider config.
+func newPluginProvider(pluginBinDir string, provider kubeletconfig.CredentialProvider) (*pluginProvider, error) {
+	mediaType := "application/json"
+	info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return nil, fmt.Errorf("unsupported media type %q", mediaType)
+	}
+
+	gv, ok := apiVersions[provider.APIVersion]
+	if !ok {
+		return nil, fmt.Errorf("invalid apiVersion: %q", provider.APIVersion)
+	}
+
+	return &pluginProvider{
+		matchImages:          provider.MatchImages,
+		cache:                cache.NewExpirationStore(cacheKeyFunc, &cacheExpirationPolicy{}),
```

Comment:
Reading through the NewExpirationStore  implementation, it looks like expired keys are only removed from this cache on an attempted get or on a list of all keys (which we don't do from pluginProvider). That means if Provide was perpetually called with images in different registries and results were cached by registry, the cache would grow without bound. Add an item to the beta graduation list to work through pruning this periodically

Reply:
> Add an item to the beta graduation list to work through pruning this periodically

ack